odoo/addons/base/tests/test_ir_attachment.py

```diff
@@ -75,53 +75,58 @@ def test_06_linked_record_permission(self):
+        vals = {'name': 'attach', 'res_id': record1.id, 'res_model': record1._name}
+        a1 = Attachment.create(vals)
 
-        # remove access to linked record a1
+        # remove all access rights to record1
         rule = self.env['ir.rule'].create({
-            'name': 'test_rule', 'domain_force': "[('id', '!=', %s)]" % a1.id,
-            'model_id': self.env.ref('base.model_ir_attachment').id,
+            'name': 'test_rule', 'domain_force': "[('id', '!=', %s)]" % record1.id,
+            'model_id': self.env['ir.model']._get_id(record1._name),
         })
-        a2.invalidate_cache(ids=a2.ids)
+        a1.flush()
+        a1.invalidate_cache(ids=a1.ids)
 
-        # no read permission on linked record
+        # no read permission on record1 -> no read permission on a1
         with self.assertRaises(AccessError):
-            a2.datas
+            a1.datas
 
-        # read permission on linked record
+        # add read permission on record1 -> one can read a1 but not modify it
         rule.perm_read = False
-        a2.datas
+        a1.datas
 
-        # no write permission on linked record
         with self.assertRaises(AccessError):
             Attachment.create(vals)
         with self.assertRaises(AccessError):
-            a2.write({'datas': self.blob2_b64})
+            a1.write({'datas': self.blob2_b64})
         with self.assertRaises(AccessError):
-            a2.unlink()
+            a1.unlink()
 
-        # can't copy with same linked record without write permission
+        # create two other records, one without write access and one with it
+        unwritable2 = env['res.users.log'].create({})
         with self.assertRaises(AccessError):
-            a2.copy()
-        # can't copy to other record without write permission
+            unwritable2.write({})
+
+        writable3 = Attachment.create({'name': 'yes'})
+        writable3.name = 'canwrite'
+
+        # cannot copy with same linked record without write permission
         with self.assertRaises(AccessError):
-            a2.copy({'res_model': log._name, 'res_id': log.id})
+            a1.copy()
+        # cannot copy to other record without write permission
+        with self.assertRaises(AccessError):
+            a1.copy({'res_model': unwritable2._name, 'res_id': unwritable2.id})
         # can copy to a record with write permission
-        a3 = Attachment.create({'name': 'yes'})
-        a3.name = 'canwrite'
-        a2_copy = a2.copy({'res_id': a3.id})
+        a3 = a1.copy({'res_model': writable3._name, 'res_id': writable3.id})
+
         # can copy to self with write permission
-        a2_copy.copy()
-        # can't copy back to record without write permission
+        a3.copy()
+        # cannot copy back to record without write permission
         with self.assertRaises(AccessError):
-            a2_copy.copy({'res_id': vals['res_id']})
+            a3.copy({'res_model': record1._name, 'res_id': record1.id})
```

Comment:
This rather misses the point, which was copying cross records without even *touching* the model, to verify that that behaves correctly.

The behaviour when the model is explicit is already checked at lines 119 and 121.